### PR TITLE
CDF Boost Removal

### PR DESCRIPTION
**Pull request by pranavsubramani**
This pull request is WIP that creates utility functions to implement the standard CDF functions without using Boost.

- incomplete beta function created, `ibeta` (Note: These are regularized, meaning that the range of the function is in `[0, 1]`)
- incomplete gamma (lower and upper) created
- Regimes where asymptotic series work better are added too
- `constant.birch` created where constants have been placed (Note: For specific implementations, constants might change, like TINY/STOP constants and this is reflected in the PR)
- `test_cdf.birch` epsilon threshold changed.

TODO:

- Test utility functions to ensure that they work in the appropriate way (potentially using boost)
- `Digamma` is not yet changed but it might be required in the future

### CODE REVIEW

File: libraries/Standard/src/math/special.birch [hunk 1]
```diff
@@ -90,36 +104,79 @@ function lgamma(x:Real32, p:Integer) -> Real32 {
  * The beta function.
  */
 function beta(x:Real64, y:Real64) -> Real64 {
-  cpp {{
-  return boost::math::beta(x, y);
-  }}
+  return (gamma(x) * gamma(y)) / (gamma(x + y));
```
Comment (devmotion) [inline libraries/Standard/src/math/special.birch:107]: This is prone to overflow and hence usually calculations are performed in log space. However, even then one has to be careful that no cancellation errors occur. See e.g. [the explanation and implementation in the boost libraries](https://www.boost.org/doc/libs/1_46_1/libs/math/doc/sf_and_dist/html/math_toolkit/special/sf_beta/beta_function.html#math_toolkit.special.sf_beta.beta_function.implementation) or the [Julia implementations](https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/src/gamma.jl#L769-L803).

File: libraries/Standard/src/math/cdf.birch [hunk 2]
```diff
@@ -152,13 +150,11 @@ function cdf_inverse_gamma_gamma(x:Real, k:Real, α:Real, β:Real) -> Real {
...
   assert 0.0 < λ;
-  
+
   if x <= 0.0 {
     return 0.0;
   } else {
-    cpp{{
-    return boost::math::cdf(boost::math::exponential_distribution<>(λ), x);
-    }}
+    return 1.0 - exp(- λ * x);
```
[the remark this answers is not in the document]
Comment (devmotion) [reply libraries/Standard/src/math/cdf.birch:157]: Maybe use `expm1` to avoid catastrophic cancellations if `-λ*x` is close to 0?
```suggestion
    return -expm1(-λ*x);
```

File: libraries/Standard/src/math/cdf.birch [hunk 3]
```diff
@@ -174,9 +170,11 @@ function cdf_exponential(x:Real, λ:Real) -> Real {
 function cdf_weibull(x:Real, k:Real, λ:Real) -> Real {
   assert 0.0 < k;
   assert 0.0 < λ;
-  cpp{{
-  return boost::math::cdf(boost::math::weibull_distribution<>(k, λ), x);
-  }}
+  if (x >= 0) {
+    return 1.0 - exp(-pow(x / λ, k));
```
[the remark this answers is not in the document]
Comment (devmotion) [reply libraries/Standard/src/math/cdf.birch:174]: Same here, maybe use `expm1`?
```suggestion
    return -expm1(-pow(x/λ, k));
```

File: libraries/Standard/src/math/cdf.birch (hunk 2 above)
[the remark this answers is not in the document]
Comment (lawmurray) [reply libraries/Standard/src/math/cdf.birch:157]: Nice catch, agreed!

File: libraries/Standard/src/math/special.birch (hunk 1 above)
Comment (pranavsubramani) [reply libraries/Standard/src/math/special.birch:107]: Thanks for this! I'll have a look at changing this implementation to follow this.

File: libraries/Standard/src/math/cdf.birch (hunk 2 above)
[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/cdf.birch:157]: Yeah, that's a great suggestion! There are a couple of other places where we might have a catastrophic cancellation issue, specifically with regards to `log1p`.

File: libraries/Standard/src/math/cdf.birch [hunk 4]
```diff
@@ -58,13 +57,12 @@ function cdf_negative_binomial(x:Integer, k:Integer, ρ:Real) -> Real {
...
-  
+
   if x < 0 {
     return 0.0;
   } else {
-    cpp{{
-    return boost::math::cdf(boost::math::poisson_distribution<>(λ), x);
-    }}
+    let k <- floor(Real(x)) + 1.0;
+    return upper_inc_gamma(k, λ);
```
[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/cdf.birch:65]: Shouldn't it be `floor(x) + 1` or am I missing something here?

File: libraries/Standard/src/test/test_cdf.birch [hunk 5]
```diff
@@ -49,9 +49,9 @@ function test_cdf(q:Distribution<Real>, N:Integer) {
     let x <- from! + (n - 0.5)*(to! - from!)/N;
     let C <- q.cdf(x)!;
     P <- P + q.pdf(x)*(to! - from!)/N;
-   
+
     let δ <- abs(C - P);
-    let ε <- 1.0/sqrt(Real(N));
+    let ε <- 5.0/sqrt(Real(N));
```
[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/test/test_cdf.birch:54]: Ah, thanks for the clarification!

File: libraries/Standard/src/math/constant.birch [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * $\infty$
+ */
+inf:Real64 <- 1.0/0.0;
+
+/**
+ * NaN.
+ */
+nan:Real64 <- 0.0/0.0;
+
+// constants from scipy
```
[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/constant.birch]: Good point, I've added to the constant a prefix to each of the constants to denote where they are used. For example:

`small -> IGAM_SMALL`

File: libraries/Standard/src/math/constant.birch [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+SMALL:Real64 <- 20;
+LARGE:Real64 <- 200;
+SMALLRATIO:Real64 <- 0.3;
+LARGERATIO:Real64 <- 4.5;
+MAXLOG:Real64 <- 709.782712893383996732;
+LANCZOS_G:Real64 <- 6.024680040776729583740234375;
+big:Real64 <- 4.503599627370496e15;
+biginv:Real64 <- 2.22044604925031308085e-16;
```
[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/constant.birch:25]: This is a great idea, I've renamed the variables and converted them to upper case :)

[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/constant.birch:25]: Re: local vs. global -- I had no idea that local variables can be accessed faster. I'll move some of the local variables back into the functions where they're called.

File: libraries/Standard/src/math/cdf.birch (hunk 4 above)
[the remark this answers is not in the document]
Comment (lawmurray) [reply libraries/Standard/src/math/cdf.birch:62]: `x` is already an integer, no?

[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/cdf.birch:62]: Sorry, that was my mistake, the correct formulation should just be:

`return upper_inc_gamma(x + 1.0, λ)`

[the remark this answers is not in the document]
Comment (pranavsubramani) [reply libraries/Standard/src/math/cdf.birch:62]: I've reflected most of the PR related changes in the most recent commits (across all the files, since there were quite a few changes to go around).

File: libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch [hunk 8]
```diff
@@ -1,8 +1,10 @@
 /*
-    Testing lower_inc_gamma
-*/
+ * Testing `lower_inc_gamma`
+ */
 
 program test_lower_inc_gamma(a:Real64, x:Real64) {
+    // note that this is not a robust test
+    // TODO(pranavs): Have a streamlined test for this function
```
Comment (lawmurray) [inline libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch:7]: We can remove this test, unless the plan is to write a comprehensive one (is there any need?)

Comment (lawmurray) [reply libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch:7]: Likewise the other tests. If there's a need to test with random arguments, or perhaps just test edge cases, we could include tests. Otherwise these functions are called as part of the cdf tests, which should reveal errors.

File: libraries/Standard/src/math/special.birch [hunk 9]
```diff
@@ -363,6 +366,10 @@ function upper_inc_gamma(a:Real, x:Real) -> Real {
 
   assert x >= Real(0.0);
```
Comment (lawmurray) [inline libraries/Standard/src/math/special.birch:367]: Still quite a few of these casts of literals around, might just want to do a search and replace to remove them :)

File: libraries/Standard/src/math/special.birch [hunk 10]
```diff
@@ -581,15 +592,15 @@ function asymptotic_series(a:Real, x:Real, func:Integer) -> Real {
   let flag1 <- true;
   let k <- 1;
   while (k <= K && flag1) {
-    ck <- asymptotic_series_d[k,1];
+    ck <- IGAM_ASYMPTOTIC_SERIES_D[k,1];
     let flag2 <- true;
     let n <- 2;
     while (n <= N && flag2) {
       if (n > maxpow) {
```
Comment (lawmurray) [inline libraries/Standard/src/math/special.birch:599]: I might be missing something here, but isn't `maxpow` initialized to `0`, and `n` to `2`, and they each increment by one on each iteration of the loop, so `n > maxpow` is always true? Is that supposed to happen or does it indicate a bug?

File: libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch (hunk 8 above)
Comment (pranavsubramani) [reply libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch:7]: I thought about this a bit, and when running some of the tests locally, I checked to make sure that all of the different regimes were tested. While I'm convinced that the cdf code stress tests these implementations, I did want to compare these against boost to make sure that things were okay.

File: libraries/Standard/src/math/special.birch (hunk 9 above)
Comment (pranavsubramani) [reply libraries/Standard/src/math/special.birch:367]: Oh, maybe I misunderstood, I thought we wanted to cast to Real explicitly for the float64 implementations and similarly for the float32 implementations to make this GPU compatible?

File: libraries/Standard/src/math/special.birch (hunk 10 above)
Comment (pranavsubramani) [reply libraries/Standard/src/math/special.birch:599]: Good catch, I took a look at the scipy implementation and it appears like this is the case in there as well: https://github.com/scipy/scipy/blob/master/scipy/special/cephes/igam.c#L395

Let me know what you think, if it is, I can open an issue in scipy, find the fix and fix it here too.

File: libraries/Standard/src/math/special.birch (hunk 9 above)
Comment (lawmurray) [reply libraries/Standard/src/math/special.birch:367]: I may not have explained very well! Let me try again:

1. There are types `Real64` and `Real32`.
2. `Real` is just an alias for `Real64` (there's a `type Real = Real64` in the standard library).
3. Literals like `0.0` are of type `Real`.
4. When we start wanting to use 32-bit (i.e. single precision) floating point, there are a couple of things we may want to do. One is to provide explicit `Real64` and `Real32` overloads of relevant functions, as for the standard math functions already (e.g. `exp(x:Real64)` and `exp(x:Real32)`). For the 32-bit overloads we will need to cast literals to `Real32` where necessary, to ensure that we are not unintentionally promoting 32-bit floating point computations to 64-bit (in Birch, as in C/C++, floats are implicitly promoted to doubles when in expressions with them). An alternative is to enable 32-bit literals in Birch, which in C/C++ are written e.g. `0.0f`.
5. The other thing we may want to do is what [LibBi](https://libbi.org) already does: provide a build option like `--enable-single` that redefines `type Real = Real32`, thus making all use of `Real` now mean 32-bit, including literals being 32-bit.

So `Real32(0.0)` is and will be a meaningful thing to do to ensure that operations are 32-bit, but `Real(0.0)` doesn't change anything---the literal `0.0` is already of type `Real`, whether `Real = Real64` or `Real = Real32`.

File: libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch (hunk 8 above)
Comment (lawmurray) [reply libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch]: Ok, that's fine. So is the plan that we temporarily include tests that compare against Boost, and remove them later (given the intent of these changes is to ultimately remove dependency on Boost)?

File: libraries/Standard/src/math/special.birch (hunk 10 above)
Comment (lawmurray) [reply libraries/Standard/src/math/special.birch:599]: Ah wait, I've misunderstood; `maxpos` is not reset to zero in each iteration of the outer loop over `k`. So it gets increased on the first iteration of the outer loop, then maintains that higher value for the remaining iterations. So there is something meaningful happening here.

File: libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch (hunk 8 above)
Comment (pranavsubramani) [reply libraries/Standard/src/test/cdf/utility/test_lower_inc_gamma.birch]: I think for now, we can test the cdf implementations and it should be fine. But maybe for added robustness, we can add tests against Boost as well in the future?

File: libraries/Standard/src/math/special.birch (hunk 9 above)
Comment (pranavsubramani) [reply libraries/Standard/src/math/special.birch]: Fixed :)

**Comment by lawmurray** [comment]
Thanks for all the efforts on this @pranavsubramani, this puts us one step closer to speeding up compile times, and eliminating some complex dependencies for GPU and other hardware support. Now merged!
